hydra-test-utils/src/Test/Hydra/Prelude.hs

```diff
@@ -137,3 +156,83 @@ newtype ReasonablySized a = ReasonablySized a
 
 instance Arbitrary a => Arbitrary (ReasonablySized a) where
   arbitrary = ReasonablySized <$> reasonablySized arbitrary
+
+-- Like 'coverTable', but construct the weight requirements generically from the
+-- label type-representation.
+--
+--     data MyCoverLabel = Case1 | Case2 deriving (Generic, Data, Show)
+--
+--     forAll arbitrary $ \(lbl :: MyCoverLabel, arg) ->
+--         myProp arg
+--         & genericCoverTable @UniformCoverage [lbl]
+--         & checkCoverage
+--
+--     ===
+--
+--     forAll arbitrary $ (lbl :: MyCoverLabel, arg) ->
+--        myProp arg
+--        & coverTable "MyCoverTable" [(Case1, 50), (Case2, 50)]
+--        & tabulate "MyCoverTable" [show lbl]
+--        & checkCoverage
+--
+genericCoverTable ::
+  forall distr lbl prop.
+  (GCoverTableRequirements distr lbl (Rep lbl), Testable prop) =>
+  [lbl] ->
+  prop ->
+  Property
+genericCoverTable labels =
+  coverTable tableName (gCoverTableRequirements @distr @lbl @(Rep lbl))
+    . tabulate tableName (show <$> labels)
+ where
+  tableName = tyConName $ typeRepTyCon $ typeRep (Proxy @lbl)
+
+-- | A Constraint alias to ease signatures on the consumer-side.
+type GUniformCoverage a =
+  GCoverTableRequirements 'UniformCoverage a (Rep a)
+
+-- | Type of distribution requirements.
+data CoverTableDistribution
+  = -- | Equal chance for all branches.
+    UniformCoverage
+
+-- | A type-class for generically creating a QuickCheck's 'coverTable'
+-- weight requirement. It is parameterized by the type of distribution wanted.
+--
+-- Any basic sum of constructors is an instance of `GCoverTableRequirements`.
+class
+  (Data t, Generic t, Show t) =>
+  GCoverTableRequirements (k :: CoverTableDistribution) t (f :: Type -> Type)
+  where
+  gCoverTableRequirements :: [(String, Double)]
+
+instance
+  (Data t, Generic t, Show t, GCoverTableRequirements k t f) =>
+  GCoverTableRequirements k t (D1 c f)
+  where
+  gCoverTableRequirements = gCoverTableRequirements @k @t @f
+
+instance
+  (Data t, Generic t, Show t, GCoverTableRequirements k t f) =>
+  GCoverTableRequirements k t (C1 c f)
+  where
+  gCoverTableRequirements = gCoverTableRequirements @k @t @f
+
+instance
+  (Data t, Generic t, Show t) =>
+  GCoverTableRequirements 'UniformCoverage t (f :+: g)
+  where
+  gCoverTableRequirements =
+    let lengthI = toInteger . length
+        cs = dataTypeConstrs (dataTypeOf (Prelude.undefined :: t))
```

Comment:
This seems to be the core idea of the generic mechanism. I ask myself whether we could use this via a simpler `uniformLabels :: Generic a => a -> [(String, Double)]` function and just pass this to `coverTable` or so?

Reply:
Hmmm. I think this question is the same as below but asked differently, so see answer on the other comment :no_mouth: